src/lib/server/map.c

```diff
@@ -799,69 +803,123 @@ static void map_sort_split(map_t *source, map_t **front, map_t **back)
 	 *	Fast advances twice as fast as slow, so when it gets to the end,
 	 *	slow will point to the middle of the linked list.
 	 */
-	slow = source;
-	fast = source->next;
+	slow = *source;
+	fast = fr_dlist_next(&list->head, slow);
 
 	while (fast) {
-		fast = fast->next;
+		fast = fr_dlist_next(&list->head, fast);
 		if (fast) {
-			slow = slow->next;
-			fast = fast->next;
+			slow = fr_dlist_next(&list->head, slow);
+			fast = fr_dlist_next(&list->head, fast);;
 		}
 	}
 
-	*front = source;
-	*back = slow->next;
-	slow->next = NULL;
+	*front = *source;
+	*back = fr_dlist_next(&list->head, slow);
+	/*
+	 * NULL terminate the first half of the list to make two lists
+	 */
+	if (slow) slow->entry.next = NULL;
 }
 
-static map_t *map_sort_merge(map_t *a, map_t *b, fr_cmp_t cmp)
+static map_t *map_sort_merge(fr_map_list_t *list, map_t **a, map_t **b, fr_cmp_t cmp)
 {
 	map_t *result = NULL;
+	map_t *next;
 
-	if (!a) return b;
-	if (!b) return a;
+	if (!a) return *b;
+	if (!b) return *a;
 
 	/*
 	 *	Compare things in the maps
 	 */
 	if (cmp(a, b) <= 0) {
-		result = a;
-		result->next = map_sort_merge(a->next, b, cmp);
+		result = *a;
+		next = fr_dlist_next(&list->head, (*a));
+		next = map_sort_merge(list, &next, b, cmp);
+		result->entry.next = &next->entry;
 	} else {
-		result = b;
-		result->next = map_sort_merge(a, b->next, cmp);
+		result = *b;
+		next = fr_dlist_next(&list->head, (*b));
+		next = map_sort_merge(list, a, &next, cmp);
+		result->entry.next = &next->entry;
 	}
 
 	return result;
 }
 
-/** Sort a linked list of #map_t using merge sort
+/** Recursive sort routine for dlist of map_t
  *
- * @param[in,out] maps List of #map_t to sort.
+ * @param[in,out] list head of list being sorted
+ * @param[in,out] map First item in current section of list being sorted.
  * @param[in] cmp to sort with
  */
-void map_sort(map_t **maps, fr_cmp_t cmp)
+static void _map_sort(fr_map_list_t *list, map_t **map, fr_cmp_t cmp)
 {
-	map_t *head = *maps;
 	map_t *a;
 	map_t *b;
 
 	/*
 	 *	If there's 0-1 elements it must already be sorted.
 	 */
-	if (!head || !head->next) {
+	if (!*map || (!(*map)->entry.next)) return;
+	map_sort_split(list, map, &a, &b);	/* Split into sublists */
+	_map_sort(list, &a, cmp);		/* Traverse left */
+	_map_sort(list, &b, cmp);		/* Traverse right */
+
+	/*
+	 *	merge the two sorted lists together
+	 */
+	*map = map_sort_merge(list, &a, &b, cmp);
+}
+
+
+/** Sort a doubly linked list of #map_t using merge sort
+ *
+ * @param[in,out] list of #map_t to sort.
+ * @param[in] cmp to sort with
+ */
+void map_sort(fr_map_list_t *list, fr_cmp_t cmp)
```

Comment:
@ndptech I have a feeling we'll probably want to sort different structures built on dlists.  Maybe add merge sort functions to dlist.h and use them here?  Shouldn't be much work seeing as you have most of the code :)

Reply:
It's now in dlist.h, and I've got that being used by fr_pair_list_sort() as well.